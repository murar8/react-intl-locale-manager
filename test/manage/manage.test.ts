import { execSync } from "child_process";
import { copySync, readJSONSync, removeSync } from "fs-extra";
import path from "path";

const BIN = path.join(__dirname, "../../bin/locale-manager");

const FIXTURES_DIR = path.join(__dirname, "fixtures");
const TEST_DIR = path.join(__dirname, "actual");

const run = (fixture: string, args: string[]) => {
  copySync(path.join(FIXTURES_DIR, fixture), TEST_DIR);
  return execSync(["node", BIN, ...args].join(" "), { cwd: TEST_DIR });
};

beforeEach(() => removeSync(TEST_DIR));

it("Should output the translation files in a directory when using the --out-dir option.", () => {
  run("base", ["-l", "en,es", "-d", "locales", "./**/*.{ts,tsx}", "./**/*.js"]);

  for (const code of ["en", "es"]) {
    expect(readJSONSync(path.join(TEST_DIR, "locales", `${code}.json`))).toMatchSnapshot();
  }
});

it("Should output the translation files in a single file when using the --out-file option.", () => {
  run("base", ["-l", "en,es", "-f", "locales.json", "./**/*.{ts,tsx}", "./**/*.js"]);

  expect(readJSONSync(path.join(TEST_DIR, "locales.json"))).toMatchSnapshot();
});

it("Should not extract messages from the paths defined in the --ignore option.", () => {
  run("base", [
    "-l",
    "en,es",
    "-f",
    "locales.json",
    "-i",
    "./*.tsx",
    "./**/*.{ts,tsx}",
    "./**/*.js",
  ]);

  expect(readJSONSync(path.join(TEST_DIR, "locales.json"))).toMatchSnapshot();
});

it("Should pass extraction options to babel-plugin-react-intl.", () => {
  run("optional", [
    "-l",
    "en,es",
    "-f",
    "locales.json",
    "--module-source-name",
    "intl",
    "--additional-component-names",
    "FormattedDummy1,FormattedDummy2",
    "--extract-from-format-message-call",
    "./**/*.{js,ts,jsx,tsx}",
  ]);

  expect(readJSONSync(path.join(TEST_DIR, "locales.json"))).toMatchSnapshot();
});

it("Should merge existing translations with the extracted messages.", () => {
  run("changes", ["-l", "en,es", "-f", "locales.json", "./**/*.{js,ts,jsx,tsx}"]);
  expect(readJSONSync(path.join(TEST_DIR, "locales.json"))).toMatchSnapshot();
});

it("Should print information about the changes to stdout.", () => {
  const stdout = run("changes", ["-l", "en,es", "-f", "locales.json", "./**/*.{js,ts,jsx,tsx}"]);
  expect(stdout.toString()).toMatchSnapshot();
});

it("Should print information about empty translation keys to stdout.", () => {
  const stdout = run("no_changes", ["-l", "en,es", "-f", "locales.json", "./**/*.{js,ts,jsx,tsx}"]);
  expect(stdout.toString()).toMatchSnapshot();
});

it("Should show the help message when invoked with the --help option.", () => {
  const stdout = run("base", ["--help"]);
  expect(stdout.toString()).toMatchSnapshot();
});
