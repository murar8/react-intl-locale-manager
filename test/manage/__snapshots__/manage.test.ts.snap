// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should merge existing translations with the extracted messages. 1`] = `
Object {
  "en": Object {
    "message1": "content 1",
    "message2": "updated 2",
    "message3": "updated 3",
    "message4": "content 4",
    "message5": "content 5",
  },
  "es": Object {
    "message1": "content 1",
    "message2": "content 2",
    "message3": "content 3",
    "message4": "content 4",
    "message5": "content 5",
  },
}
`;

exports[`Should not extract messages from the paths defined in the --ignore option. 1`] = `
Object {
  "en": Object {
    "message4": "content 4",
    "message5": "",
  },
  "es": Object {
    "message4": "content 4",
    "message5": "",
  },
}
`;

exports[`Should output the translation files in a directory when using the --out-dir option. 1`] = `
Object {
  "message1": "content 1",
  "message2": "",
  "message3": "content 3",
  "message4": "content 4",
  "message5": "",
}
`;

exports[`Should output the translation files in a directory when using the --out-dir option. 2`] = `
Object {
  "message1": "content 1",
  "message2": "",
  "message3": "content 3",
  "message4": "content 4",
  "message5": "",
}
`;

exports[`Should output the translation files in a single file when using the --out-file option. 1`] = `
Object {
  "en": Object {
    "message1": "content 1",
    "message2": "",
    "message3": "content 3",
    "message4": "content 4",
    "message5": "",
  },
  "es": Object {
    "message1": "content 1",
    "message2": "",
    "message3": "content 3",
    "message4": "content 4",
    "message5": "",
  },
}
`;

exports[`Should pass extraction options to babel-plugin-react-intl. 1`] = `
Object {
  "en": Object {
    "message1": "",
    "message2": "message 2",
    "message3": "",
  },
  "es": Object {
    "message1": "",
    "message2": "message 2",
    "message3": "",
  },
}
`;

exports[`Should print information about the changes to console. 1`] = `
" Found 1 Duplicate ids: 
  ID: message3
    at a.js:7:12
    at b.js:5:12

 Added 5 IDs: 
  message1
  message2
  message3
  message4
  message5

 Removed 2 IDs: 
  message6
  message7

 Added 1 locales: 
  es

 Removed 2 locales: 
  de
  fr

"
`;

exports[`Should show the help message when invoked with the --help option. 1`] = `
"Usage: locale-manager [options] [files..]

Positionals:
  files  Files to be seached for translations. Every glob pattern in this list will be expanded.  [string]

Options:
  --help                              Show help  [boolean]
  --version                           Show version number  [boolean]
  -l, --languages                     Comma-separated list of language codes to support. A translation will be generated and mantained for every code in this list.  [string]
  -f, --out-file                      Path to the file where the extracted messages will be output in a single JSON object grouped by locale.  [string]
  -d, --out-dir                       Path to the directory where the extracted messages will be output generating a [locale].json file for each locale.  [string]
  -i, --ignore                        Glob pattern designationg the files to exlcude from the translation process.To define more than one ignore pattern, just list the flag multiple times.  [string]
  --module-source-name                The ES6 module source name of the React Intl package.  [string]
  --additional-component-names        Comma separated list of component names to extract messages from. Note that default we check for the fact that 'FormattedMessage' is imported from '--module-source-name' to make sure variable alias works. This option does not do that so it's less safe.,  [string]
  --extract-from-format-message-call  Opt-in to extract from 'intl.formatMessage' calls with the restriction that it has to be called with an object literal such as 'intl.formatMessage({ id: 'foo', ...})  [boolean]

Examples:
    $0 -l en,es -d ./locales -i src/**/*.test.js -i src/**/*.spec.js  src/**/*.js
    $0 --extract-from-format-message-call -l en,es -f locales.json src/**/*.tsx

For additional information, visit: https://github.com/murar8/react-intl-locale-manager
"
`;
